src/c++/perf_analyzer/test_request_rate_manager.cc

```diff
@@ -664,121 +811,311 @@ TEST_CASE(
+   ]})"};
+
+  model_tensor1.is_shape_tensor_ = true;
+  model_tensor2.is_optional_ = true;
+
+  SUBCASE("batch 1")
+  {
+    params.batch_size = 1;
+    expected_results = {{1, 21}, {1, 22}, {1, 23}, {1, 21}};
+  }
+  SUBCASE("batch 2")
   {
-    is_sequence_model = false;
-    ParameterizeAsyncAndStreaming(params.async, params.streaming);
+    params.batch_size = 2;
+    expected_results = {{1, 21, 22}, {1, 23, 21}, {1, 22, 23}, {1, 21, 22}};
   }
-  SUBCASE("sequence")
+  SUBCASE("batch 4")
   {
-    is_sequence_model = true;
-    params.num_of_sequences = 1;
-    ParameterizeAsyncAndStreaming(params.async, params.streaming);
+    params.batch_size = 4;
+    expected_results = {{1, 21, 22, 23, 21},
+                        {1, 22, 23, 21, 22},
+                        {1, 23, 21, 22, 23},
+                        {1, 21, 22, 23, 21}};
   }
 
-  std::shared_ptr<MockModelParser> mmp{
-      std::make_shared<MockModelParser>(is_sequence_model, false)};
-  ModelTensor model_tensor{};
-  model_tensor.datatype_ = "INT32";
-  model_tensor.is_optional_ = false;
-  model_tensor.is_shape_tensor_ = false;
-  model_tensor.name_ = "INPUT0";
-  model_tensor.shape_ = {1};
-  mmp->inputs_ = std::make_shared<ModelTensorMap>();
-  (*mmp->inputs_)[model_tensor.name_] = model_tensor;
-
-  std::shared_ptr<MockDataLoader> mdl{std::make_shared<MockDataLoader>()};
-  const std::string json_str{R"(
-{
-  "data": [
-    {
-      "INPUT0": [2000000000]
-    },
-    {
-      "INPUT0": [2000000001]
-    }
-  ]
+  TestRequestRateManager trrm(params, is_sequence_model);
+
+  tensors.push_back(model_tensor1);
+  tensors.push_back(model_tensor2);
+
+  trrm.TestCustomData(
+      tensors, json_str, expected_results, expect_init_failure,
+      expect_thread_failure);
 }
-    )"};
-  mdl->ReadDataFromStr(json_str, mmp->Inputs(), mmp->Outputs());
+
+TEST_CASE("custom_json_data: handling invalid is_shape_tensor")
+{
+  PerfAnalyzerParameters params{};
+  params.user_data = {"fake_file.json"};
+  bool is_sequence_model{false};
+
+  std::vector<std::vector<int32_t>> expected_results;
+  std::vector<ModelTensor> tensors;
+  bool expect_init_failure = false;
+  bool expect_thread_failure = false;
+
+  ModelTensor model_tensor1{};
+  model_tensor1.datatype_ = "INT32";
+  model_tensor1.is_optional_ = false;
+  model_tensor1.is_shape_tensor_ = false;
+  model_tensor1.name_ = "INPUT1";
+  model_tensor1.shape_ = {1};
+
+  ModelTensor model_tensor2 = model_tensor1;
+  model_tensor2.name_ = "INPUT2";
+
+  std::string json_str{R"({
+   "data": [
+     { "INPUT1": [1], "INPUT2": [21] },
+     { "INPUT1": [2], "INPUT2": [22] },
+     { "INPUT1": [3], "INPUT2": [23] }     
+   ]})"};
+
+
+  SUBCASE("no batching is ok")
+  {
+    model_tensor1.is_shape_tensor_ = true;
+    params.batch_size = 1;
+    expected_results = {{1, 21}, {2, 22}, {3, 23}, {1, 21}};
+  }
+  SUBCASE("batching - no shm")
+  {
+    // Currently shm and non-shm both fail for batching, but at different points
```

Comment:
Why?

Reply:
Due to the format of the existing code.
shared memory code pre-populates all of the data into shared memory in init. Thus, when InitManager() is called, it will evaluate the validity of the data and would throw an error at that point.
non-shared-memory code will evaluate the validity of the data on each sent request. Thus, it will fail in the thread.

Note that this will change so that they both fail in init after I make the overhead optimization where non-shared-memory pre-populates all of the data. I have tagged these lines with the TMA-765 tag to make sure it is cleaned up properly after my changes.